File: internal/registry/registry.go
// Package registry is the single source of truth for the hash algorithms the
// tool can compute, including their display labels and security categories.
package registry

import "hash"

// Category groups algorithms by their intended use.
type Category string

// Algorithm categories reported by Spec.Category.
const (
	CategoryCryptographic Category = "cryptographic"
	CategoryLegacy        Category = "legacy"
	CategoryChecksum      Category = "checksum"
	CategoryNonCrypto     Category = "non-cryptographic"
)

// Spec describes one algorithm. Name is the lower-case registry key; New
// returns a fresh hash.Hash for each file so digests never share state.
type Spec struct {
	Name     string
	Label    string
	Category Category
	Secure   bool
	New      func() hash.Hash
}

var specs = []Spec{
	{Name: "sha224", Label: "SHA-224", Category: CategoryCryptographic, Secure: true, New: newHashSHA224},
	{Name: "sha256", Label: "SHA-256", Category: CategoryCryptographic, Secure: true, New: newHashSHA256},
	{Name: "sha384", Label: "SHA-384", Category: CategoryCryptographic, Secure: true, New: newHashSHA384},
	{Name: "sha512", Label: "SHA-512", Category: CategoryCryptographic, Secure: true, New: newHashSHA512},
	{Name: "sha512-224", Label: "SHA-512/224", Category: CategoryCryptographic, Secure: true, New: newHashSHA512224},
	{Name: "sha512-256", Label: "SHA-512/256", Category: CategoryCryptographic, Secure: true, New: newHashSHA512256},
	{Name: "md5", Label: "MD5", Category: CategoryLegacy, Secure: false, New: newHashMD5},
	{Name: "sha1", Label: "SHA-1", Category: CategoryLegacy, Secure: false, New: newHashSHA1},
	{Name: "adler32", Label: "Adler-32", Category: CategoryChecksum, Secure: false, New: newAdler32},
	{Name: "crc32-ieee", Label: "CRC-32/IEEE", Category: CategoryChecksum, Secure: false, New: newCRC32IEEE},
	{Name: "crc32-castagnoli", Label: "CRC-32C/Castagnoli", Category: CategoryChecksum, Secure: false, New: newCRC32Castagnoli},
	{Name: "crc32-koopman", Label: "CRC-32/Koopman", Category: CategoryChecksum, Secure: false, New: newCRC32Koopman},
	{Name: "crc64-ecma", Label: "CRC-64/ECMA", Category: CategoryChecksum, Secure: false, New: newCRC64ECMA},
	{Name: "crc64-iso", Label: "CRC-64/ISO", Category: CategoryChecksum, Secure: false, New: newCRC64ISO},
	{Name: "fnv1-32", Label: "FNV-1 32", Category: CategoryNonCrypto, Secure: false, New: newFNV132},
	{Name: "fnv1a-32", Label: "FNV-1a 32", Category: CategoryNonCrypto, Secure: false, New: newFNV1a32},
	{Name: "fnv1-64", Label: "FNV-1 64", Category: CategoryNonCrypto, Secure: false, New: newFNV164},
	{Name: "fnv1a-64", Label: "FNV-1a 64", Category: CategoryNonCrypto, Secure: false, New: newFNV1a64},
	{Name: "fnv1-128", Label: "FNV-1 128", Category: CategoryNonCrypto, Secure: false, New: newFNV1128},
	{Name: "fnv1a-128", Label: "FNV-1a 128", Category: CategoryNonCrypto, Secure: false, New: newFNV1a128},
}

var byName = buildIndex()

// Algorithms returns a copy of every registered algorithm, in display order.
func Algorithms() []Spec {
	return append([]Spec(nil), specs...)
}

// Lookup returns the algorithm registered under name.
func Lookup(name string) (Spec, bool) {
	spec, ok := byName[name]
	return spec, ok
}

func buildIndex() map[string]Spec {
	index := make(map[string]Spec, len(specs))
	for _, spec := range specs {
		index[spec.Name] = spec
	}
	return index
}

File: src/core/registry.cpp
#include "core/registry.h"

#include <QHash>

namespace hash_core {
namespace {

QVector<AlgorithmSpec> buildSpecs() {
    return {
        {QStringLiteral("sha224"), QStringLiteral("SHA-224"), QStringLiteral(CategoryCryptographic),
         true, []() { return makeQtDigester(QCryptographicHash::Sha224); }},
        {QStringLiteral("sha256"), QStringLiteral("SHA-256"), QStringLiteral(CategoryCryptographic),
         true, []() { return makeQtDigester(QCryptographicHash::Sha256); }},
        {QStringLiteral("sha384"), QStringLiteral("SHA-384"), QStringLiteral(CategoryCryptographic),
         true, []() { return makeQtDigester(QCryptographicHash::Sha384); }},
        {QStringLiteral("sha512"), QStringLiteral("SHA-512"), QStringLiteral(CategoryCryptographic),
         true, []() { return makeQtDigester(QCryptographicHash::Sha512); }},
        {QStringLiteral("sha512-224"), QStringLiteral("SHA-512/224"),
         QStringLiteral(CategoryCryptographic), true, []() { return makeSha512_224(); }},
        {QStringLiteral("sha512-256"), QStringLiteral("SHA-512/256"),
         QStringLiteral(CategoryCryptographic), true, []() { return makeSha512_256(); }},
        {QStringLiteral("md5"), QStringLiteral("MD5"), QStringLiteral(CategoryLegacy), false,
         []() { return makeQtDigester(QCryptographicHash::Md5); }},
        {QStringLiteral("sha1"), QStringLiteral("SHA-1"), QStringLiteral(CategoryLegacy), false,
         []() { return makeQtDigester(QCryptographicHash::Sha1); }},
        {QStringLiteral("adler32"), QStringLiteral("Adler-32"), QStringLiteral(CategoryChecksum),
         false, []() { return makeAdler32(); }},
        {QStringLiteral("crc32-ieee"), QStringLiteral("CRC-32/IEEE"),
         QStringLiteral(CategoryChecksum), false, []() { return makeCrc32(0xEDB88320ULL); }},
        {QStringLiteral("crc32-castagnoli"), QStringLiteral("CRC-32C/Castagnoli"),
         QStringLiteral(CategoryChecksum), false, []() { return makeCrc32(0x82F63B78ULL); }},
        {QStringLiteral("crc32-koopman"), QStringLiteral("CRC-32/Koopman"),
         QStringLiteral(CategoryChecksum), false, []() { return makeCrc32(0xEB31D82EULL); }},
        {QStringLiteral("crc64-ecma"), QStringLiteral("CRC-64/ECMA"),
         QStringLiteral(CategoryChecksum), false,
         []() { return makeCrc64(0xC96C5795D7870F42ULL); }},
        {QStringLiteral("crc64-iso"), QStringLiteral("CRC-64/ISO"),
         QStringLiteral(CategoryChecksum), false,
         []() { return makeCrc64(0xD800000000000000ULL); }},
        {QStringLiteral("fnv1-32"), QStringLiteral("FNV-1 32"), QStringLiteral(CategoryNonCrypto),
         false, []() { return makeFnv1_32(); }},
        {QStringLiteral("fnv1a-32"), QStringLiteral("FNV-1a 32"), QStringLiteral(CategoryNonCrypto),
         false, []() { return makeFnv1a_32(); }},
        {QStringLiteral("fnv1-64"), QStringLiteral("FNV-1 64"), QStringLiteral(CategoryNonCrypto),
         false, []() { return makeFnv1_64(); }},
        {QStringLiteral("fnv1a-64"), QStringLiteral("FNV-1a 64"), QStringLiteral(CategoryNonCrypto),
         false, []() { return makeFnv1a_64(); }},
        {QStringLiteral("fnv1-128"), QStringLiteral("FNV-1 128"), QStringLiteral(CategoryNonCrypto),
         false, []() { return makeFnv1_128(); }},
        {QStringLiteral("fnv1a-128"), QStringLiteral("FNV-1a 128"),
         QStringLiteral(CategoryNonCrypto), false, []() { return makeFnv1a_128(); }},
    };
}

} // namespace

QVector<AlgorithmSpec> algorithms() {
    static const QVector<AlgorithmSpec> specs = buildSpecs();
    return specs;
}

bool lookupAlgorithm(const QString &name, AlgorithmSpec *spec) {
    static const QHash<QString, AlgorithmSpec> index = []() {
        QHash<QString, AlgorithmSpec> built;
        built.reserve(24);
        for (const AlgorithmSpec &entry : algorithms()) {
            built.insert(entry.name, entry);
        }
        return built;
    }();
    const auto found = index.constFind(name);
    if (found == index.constEnd()) {
        return false;
    }
    if (spec) {
        *spec = found.value();
    }
    return true;
}

} // namespace hash_core
